pkg/providers/amifamily/ami.go

```diff
@@ -166,22 +166,36 @@ func (p *DefaultProvider) amis(ctx context.Context, queries []DescribeImageQuery
 				// Each image may have multiple associated sets of requirements. For example, an image may be compatible with Neuron instances
 				// and GPU instances. In that case, we'll have a set of requirements for each, and will create one "image" for each.
 				for _, reqs := range query.RequirementsForImageWithArchitecture(lo.FromPtr(image.ImageId), arch) {
-					// If we already have an image with the same set of requirements, but this image is newer, replace the previous image.
+					// Checks and store for AMIs
+					// Following checks are needed in order to always priortize non deprecated AMIs
+					// If we already have an image with the same set of requirements, but this image (candidate) is newer, replace the previous (existing) image.
+					// If we already have an image with the same set of requirements which is deprecated, but this image (candidate) is newer or non deprecated, replace the previous (existing) image
 					reqsHash := lo.Must(hashstructure.Hash(reqs.NodeSelectorRequirements(), hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true}))
+					candidateDeprecated := parseTimeWithDefault(lo.FromPtr(image.DeprecationTime), maxTime).Unix() <= time.Now().Unix()
 					if v, ok := images[reqsHash]; ok {
-						candidateCreationTime, _ := time.Parse(time.RFC3339, lo.FromPtr(image.CreationDate))
-						existingCreationTime, _ := time.Parse(time.RFC3339, v.CreationDate)
-						if existingCreationTime == candidateCreationTime && lo.FromPtr(image.Name) < v.Name {
+						existingCreationTime := parseTimeWithDefault(v.CreationDate, minTime)
+						candidateCreationTime := parseTimeWithDefault(lo.FromPtr(image.CreationDate), minTime)
+						// If both AMIs have the same creation time
+						// And if the existing AMI is non-deprecated and the candidate AMI is deprecated, return the existing AMI
+						if existingCreationTime == candidateCreationTime && !v.Deprecated && candidateDeprecated && lo.FromPtr(image.Name) < v.Name {
```

Comment:
consider implementing this in a similar way to how strings.Compare() is implemented, so that we can do compare(a ami, b ami) == -1/0/1 and check that way. would simplify the overall code from L176 to L191

Reply:
1. if both aren't deprecated, return newer one
2. if one is deprecated, return the non-deprecated
3. if both are deprecated, return the newer one

Reply:
@shabbskagalwala Did you take a look at this comment?